homeassistant/components/habitica/config_flow.py

```diff
@@ -0,0 +1,107 @@
+
+from .const import (  # pylint: disable=unused-import
+    API_STATUS,
+    API_USER_PROFILE,
+    CONF_API_USER,
+    DEFAULT_URL,
+    DOMAIN,
+)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_USER): str,
+        vol.Required(CONF_API_KEY): str,
+        vol.Optional(CONF_NAME): str,
+        vol.Optional(CONF_URL, default=DEFAULT_URL): str,
+    }
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect."""
+
+    def check_api_status(url=DEFAULT_URL) -> bool:
+        request = requests.get(API_STATUS)
+        if request.status_code == HTTP_OK:
+            return True
+        return False
+
+    def authenticate(username, password, url=DEFAULT_URL) -> bool:
+        auth_headers = {"x-api-user": username, "x-api-key": password}
+        request = requests.get(API_USER_PROFILE, headers=auth_headers)
+        if request.status_code == HTTP_OK:
+            return True
+        return False
+
+    is_valid = await hass.async_add_executor_job(check_api_status, data[CONF_URL])
+    if is_valid is False:
+        raise CannotConnect()
+
+    is_valid = await hass.async_add_executor_job(
+        authenticate, data[CONF_API_USER], data[CONF_API_KEY], data[CONF_URL]
+    )
+    if is_valid is False:
+        raise InvalidAuth()
+    return {"title": f"{data.get('name', 'Default username')}"}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for habitica."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+
+        if not user_input:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={},
+                description_placeholders={},
+            )
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "connection_error"},
+                description_placeholders={},
+            )
+        except InvalidAuth:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "invalid_credentials"},
+                description_placeholders={},
+            )
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "unknown"},
+                description_placeholders={},
+            )
+        return self.async_create_entry(title=info["title"], data=user_input)
```

Comment:
We should check that we don't create more than one entry for the same habitica instance.

Reply:
I believe setting a unique ID per cofig must be sufficient, right?

Reply:
Yes, if there's a unique id, that's the best:
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

Note that the unique id must both be unique and stable. We don't allow ip address eg.

Reply:
Sweet. I ll just user API_USER_ID, which is unique per user and is always present